/**
 * NCBI Genetic Code Tables
 *
 * Complete implementation of all 31 NCBI genetic codes for translation
 * Reference: https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi
 *
 * Each genetic code defines how DNA/RNA codons map to amino acids
 * Different organisms use different genetic codes (e.g., mitochondrial vs nuclear)
 */

import { reverseComplement } from './sequence-manipulation';

/**
 * Genetic code identifiers matching NCBI standards
 */
export enum GeneticCode {
  STANDARD = 1,
  VERTEBRATE_MITOCHONDRIAL = 2,
  YEAST_MITOCHONDRIAL = 3,
  MOLD_MITOCHONDRIAL = 4,
  INVERTEBRATE_MITOCHONDRIAL = 5,
  CILIATE_NUCLEAR = 6,
  ECHINODERM_MITOCHONDRIAL = 9,
  EUPLOTID_NUCLEAR = 10,
  BACTERIAL_PLASTID = 11,
  ALTERNATIVE_YEAST_NUCLEAR = 12,
  ASCIDIAN_MITOCHONDRIAL = 13,
  ALTERNATIVE_FLATWORM_MITOCHONDRIAL = 14,
  CHLOROPHYCEAN_MITOCHONDRIAL = 16,
  TREMATODE_MITOCHONDRIAL = 21,
  SCENEDESMUS_MITOCHONDRIAL = 22,
  THRAUSTOCHYTRIUM_MITOCHONDRIAL = 23,
  RHABDOPLEURIDAE_MITOCHONDRIAL = 24,
  CANDIDATE_DIVISION_SR1 = 25,
  PACHYSOLEN_NUCLEAR = 26,
  KARYORELICT_NUCLEAR = 27,
  CONDYLOSTOMA_NUCLEAR = 28,
  MESODINIUM_NUCLEAR = 29,
  PERITRICH_NUCLEAR = 30,
  BLASTOCRITHIDIA_NUCLEAR = 31,
  BALANOPHORACEAE_PLASTID = 32,
  CEPHALODISCIDAE_MITOCHONDRIAL = 33,
}

/**
 * Codon to amino acid mapping for a genetic code
 */
interface CodonTable {
  readonly [codon: string]: string;
}

/**
 * Complete genetic code definition
 */
interface GeneticCodeDefinition {
  readonly id: number;
  readonly name: string;
  readonly shortName: string;
  readonly codons: CodonTable;
  readonly startCodons: readonly string[];
}

/**
 * Translation manager for all NCBI genetic codes
 *
 * This is intentionally a static class because it manages a singleton dataset
 * of immutable genetic code tables. The methods and data are tightly coupled,
 * and there's no meaningful instance state.
 *
 * @remarks
 * The static class pattern is appropriate here because:
 * - Genetic codes are universal biological constants
 * - The ~2000 lines of codon mappings should be loaded once
 * - Methods directly operate on the private static data
 * - No instance would ever have different genetic codes
 */
// biome-ignore lint/complexity/noStaticOnlyClass: <explanation>
export class GeneticCodeTable {
  /**
   * Base codon mappings (shared by many codes)
   * This is the standard genetic code (NCBI #1)
   */
  private static readonly STANDARD_CODE: CodonTable = {
    TTT: 'F',
    TTC: 'F',
    TTA: 'L',
    TTG: 'L',
    TCT: 'S',
    TCC: 'S',
    TCA: 'S',
    TCG: 'S',
    TAT: 'Y',
    TAC: 'Y',
    TAA: '*',
    TAG: '*',
    TGT: 'C',
    TGC: 'C',
    TGA: '*',
    TGG: 'W',
    CTT: 'L',
    CTC: 'L',
    CTA: 'L',
    CTG: 'L',
    CCT: 'P',
    CCC: 'P',
    CCA: 'P',
    CCG: 'P',
    CAT: 'H',
    CAC: 'H',
    CAA: 'Q',
    CAG: 'Q',
    CGT: 'R',
    CGC: 'R',
    CGA: 'R',
    CGG: 'R',
    ATT: 'I',
    ATC: 'I',
    ATA: 'I',
    ATG: 'M',
    ACT: 'T',
    ACC: 'T',
    ACA: 'T',
    ACG: 'T',
    AAT: 'N',
    AAC: 'N',
    AAA: 'K',
    AAG: 'K',
    AGT: 'S',
    AGC: 'S',
    AGA: 'R',
    AGG: 'R',
    GTT: 'V',
    GTC: 'V',
    GTA: 'V',
    GTG: 'V',
    GCT: 'A',
    GCC: 'A',
    GCA: 'A',
    GCG: 'A',
    GAT: 'D',
    GAC: 'D',
    GAA: 'E',
    GAG: 'E',
    GGT: 'G',
    GGC: 'G',
    GGA: 'G',
    GGG: 'G',
  };

  /**
   * All 31 NCBI genetic codes
   * Differences from standard code are applied as overrides
   */
  private static readonly GENETIC_CODES: Map<number, GeneticCodeDefinition> = new Map([
    // 1. Standard Code
    [
      1,
      {
        id: 1,
        name: 'Standard',
        shortName: 'SGC0',
        codons: GeneticCodeTable.STANDARD_CODE,
        startCodons: ['TTG', 'CTG', 'ATG'],
      },
    ],

    // 2. Vertebrate Mitochondrial
    [
      2,
      {
        id: 2,
        name: 'Vertebrate Mitochondrial',
        shortName: 'VMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AGA: '*',
          AGG: '*', // Stop instead of R
          ATA: 'M', // M instead of I
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATT', 'ATC', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 3. Yeast Mitochondrial
    [
      3,
      {
        id: 3,
        name: 'Yeast Mitochondrial',
        shortName: 'YMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          ATA: 'M', // M instead of I
          CTT: 'T',
          CTC: 'T',
          CTA: 'T',
          CTG: 'T', // T instead of L
          TGA: 'W', // W instead of stop
          CGA: '*',
          CGC: '*', // Stops instead of R
        },
        startCodons: ['ATA', 'ATG'],
      },
    ],

    // 4. Mold/Protozoan/Coelenterate Mitochondrial
    [
      4,
      {
        id: 4,
        name: 'Mold Mitochondrial',
        shortName: 'MMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TGA: 'W', // W instead of stop
        },
        startCodons: ['TTA', 'TTG', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 5. Invertebrate Mitochondrial
    [
      5,
      {
        id: 5,
        name: 'Invertebrate Mitochondrial',
        shortName: 'IMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AGA: 'S',
          AGG: 'S', // S instead of R
          ATA: 'M', // M instead of I
          TGA: 'W', // W instead of stop
        },
        startCodons: ['TTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 6. Ciliate Nuclear
    [
      6,
      {
        id: 6,
        name: 'Ciliate Nuclear',
        shortName: 'CNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'Q',
          TAG: 'Q', // Q instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 9. Echinoderm Mitochondrial
    [
      9,
      {
        id: 9,
        name: 'Echinoderm Mitochondrial',
        shortName: 'EMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AAA: 'N', // N instead of K
          AGA: 'S',
          AGG: 'S', // S instead of R
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATG', 'GTG'],
      },
    ],

    // 10. Euplotid Nuclear
    [
      10,
      {
        id: 10,
        name: 'Euplotid Nuclear',
        shortName: 'ENu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TGA: 'C', // C instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 11. Bacterial and Plant Plastid
    [
      11,
      {
        id: 11,
        name: 'Bacterial and Plant Plastid',
        shortName: 'BPl',
        codons: GeneticCodeTable.STANDARD_CODE, // Same as standard
        startCodons: ['TTG', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 12. Alternative Yeast Nuclear
    [
      12,
      {
        id: 12,
        name: 'Alternative Yeast Nuclear',
        shortName: 'AYN',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          CTG: 'S', // S instead of L
        },
        startCodons: ['CTG', 'ATG'],
      },
    ],

    // 13. Ascidian Mitochondrial
    [
      13,
      {
        id: 13,
        name: 'Ascidian Mitochondrial',
        shortName: 'AMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AGA: 'G',
          AGG: 'G', // G instead of R
          ATA: 'M', // M instead of I
          TGA: 'W', // W instead of stop
        },
        startCodons: ['TTG', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 14. Alternative Flatworm Mitochondrial
    [
      14,
      {
        id: 14,
        name: 'Alternative Flatworm Mitochondrial',
        shortName: 'AFMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AAA: 'N', // N instead of K
          AGA: 'S',
          AGG: 'S', // S instead of R
          TAA: 'Y', // Y instead of stop
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 16. Chlorophycean Mitochondrial
    [
      16,
      {
        id: 16,
        name: 'Chlorophycean Mitochondrial',
        shortName: 'CMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAG: 'L', // L instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 21. Trematode Mitochondrial
    [
      21,
      {
        id: 21,
        name: 'Trematode Mitochondrial',
        shortName: 'TMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TGA: 'W', // W instead of stop
          ATA: 'M', // M instead of I
          AGA: 'S',
          AGG: 'S', // S instead of R
          AAA: 'N', // N instead of K
        },
        startCodons: ['ATG', 'GTG'],
      },
    ],

    // 22. Scenedesmus obliquus Mitochondrial
    [
      22,
      {
        id: 22,
        name: 'Scenedesmus obliquus Mitochondrial',
        shortName: 'SoMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TCA: '*', // Stop instead of S
          TAG: 'L', // L instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 23. Thraustochytrium Mitochondrial
    [
      23,
      {
        id: 23,
        name: 'Thraustochytrium Mitochondrial',
        shortName: 'ThMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TTA: '*', // Stop instead of L
        },
        startCodons: ['ATT', 'ATG', 'GTG'],
      },
    ],

    // Remaining codes follow similar pattern...
    // I'll add the rest for completeness

    // 24. Rhabdopleuridae Mitochondrial
    [
      24,
      {
        id: 24,
        name: 'Rhabdopleuridae Mitochondrial',
        shortName: 'RMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AGA: 'S',
          AGG: 'K', // S and K instead of R
          TGA: 'W', // W instead of stop
        },
        startCodons: ['TTG', 'CTG', 'ATG', 'GTG'],
      },
    ],

    // 25. Candidate Division SR1 and Gracilibacteria
    [
      25,
      {
        id: 25,
        name: 'Candidate Division SR1',
        shortName: 'SR1',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TGA: 'G', // G instead of stop
        },
        startCodons: ['TTG', 'ATG', 'GTG'],
      },
    ],

    // 26. Pachysolen tannophilus Nuclear
    [
      26,
      {
        id: 26,
        name: 'Pachysolen tannophilus Nuclear',
        shortName: 'PtN',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          CTG: 'A', // A instead of L
        },
        startCodons: ['CTG', 'ATG'],
      },
    ],

    // 27. Karyorelict Nuclear
    [
      27,
      {
        id: 27,
        name: 'Karyorelict Nuclear',
        shortName: 'KNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'Q',
          TAG: 'Q', // Q instead of stop
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 28. Condylostoma Nuclear
    [
      28,
      {
        id: 28,
        name: 'Condylostoma Nuclear',
        shortName: 'CoNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'Q',
          TAG: 'Q', // Q instead of stop
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 29. Mesodinium Nuclear
    [
      29,
      {
        id: 29,
        name: 'Mesodinium Nuclear',
        shortName: 'MeNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'Y',
          TAG: 'Y', // Y instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 30. Peritrich Nuclear
    [
      30,
      {
        id: 30,
        name: 'Peritrich Nuclear',
        shortName: 'PeNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'E',
          TAG: 'E', // E instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 31. Blastocrithidia Nuclear
    [
      31,
      {
        id: 31,
        name: 'Blastocrithidia Nuclear',
        shortName: 'BNu',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAA: 'E',
          TAG: 'E', // E instead of stop
          TGA: 'W', // W instead of stop
        },
        startCodons: ['ATG'],
      },
    ],

    // 32. Balanophoraceae Plastid
    [
      32,
      {
        id: 32,
        name: 'Balanophoraceae Plastid',
        shortName: 'BaPl',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          TAG: 'W', // W instead of stop
        },
        startCodons: ['TTG', 'CTG', 'ATT', 'ATC', 'ATA', 'ATG', 'GTG'],
      },
    ],

    // 33. Cephalodiscidae Mitochondrial
    [
      33,
      {
        id: 33,
        name: 'Cephalodiscidae Mitochondrial',
        shortName: 'CeMt',
        codons: {
          ...GeneticCodeTable.STANDARD_CODE,
          AGA: 'S', // S instead of R
          AGG: 'K', // K instead of R
          TAA: 'Y', // Y instead of stop
          TGA: 'W', // W instead of stop
        },
        startCodons: ['TTG', 'CTG', 'ATG', 'GTG'],
      },
    ],
  ]);

  /**
   * Translate DNA sequence to amino acids
   *
   * @example
   * ```typescript
   * const protein = GeneticCodeTable.translate(
   *   'ATGGGATCC',
   *   GeneticCode.STANDARD
   * );
   * console.log(protein); // 'MGS'
   * ```
   *
   * @param sequence - DNA sequence (will convert U to T)
   * @param codeId - Genetic code to use (default: standard)
   * @param frame - Reading frame (0, 1, or 2)
   * @returns Amino acid sequence
   *
   * ðŸ”¥ ZIG CRITICAL: Translation is heavily used and should be optimized
   */
  static translate(
    sequence: string,
    codeId: GeneticCode = GeneticCode.STANDARD,
    frame: 0 | 1 | 2 = 0
  ): string {
    // Tiger Style: Assert inputs
    if (!sequence || typeof sequence !== 'string') {
      throw new Error('Sequence must be a non-empty string');
    }
    if (frame < 0 || frame > 2) {
      throw new Error('Frame must be 0, 1, or 2');
    }

    const code = GeneticCodeTable.GENETIC_CODES.get(codeId);
    if (!code) {
      throw new Error(`Unknown genetic code: ${codeId}`);
    }

    // Prepare sequence - uppercase and convert RNA to DNA
    const dna = sequence.toUpperCase().replace(/U/g, 'T');

    // Start from frame position
    let protein = '';
    for (let i = frame; i + 2 < dna.length; i += 3) {
      const codon = dna.substring(i, i + 3);
      const aa = code.codons[codon];
      protein += aa !== undefined && aa !== null && aa !== '' ? aa : 'X'; // X for unknown codons
    }

    return protein;
  }

  /**
   * Translate in all six frames (3 forward, 3 reverse)
   *
   * @param sequence - DNA sequence
   * @param codeId - Genetic code to use
   * @returns Object with all six frame translations
   */
  static translateSixFrames(
    sequence: string,
    codeId: GeneticCode = GeneticCode.STANDARD
  ): Record<string, string> {
    // Tiger Style: Assert input
    if (!sequence || typeof sequence !== 'string') {
      throw new Error('Sequence must be a non-empty string');
    }

    // Get reverse complement
    const revComp = reverseComplement(sequence);

    return {
      '+1': GeneticCodeTable.translate(sequence, codeId, 0),
      '+2': GeneticCodeTable.translate(sequence, codeId, 1),
      '+3': GeneticCodeTable.translate(sequence, codeId, 2),
      '-1': GeneticCodeTable.translate(revComp, codeId, 0),
      '-2': GeneticCodeTable.translate(revComp, codeId, 1),
      '-3': GeneticCodeTable.translate(revComp, codeId, 2),
    };
  }

  /**
   * Find all open reading frames (ORFs)
   *
   * @param sequence - DNA sequence
   * @param codeId - Genetic code to use
   * @param minLength - Minimum ORF length in amino acids (default: 20)
   * @returns Array of ORF objects with position and translation
   */
  static findORFs(
    sequence: string,
    codeId: GeneticCode = GeneticCode.STANDARD,
    minLength: number = 20
  ): Array<{
    start: number;
    end: number;
    frame: number;
    strand: '+' | '-';
    length: number;
    protein: string;
  }> {
    // Tiger Style: Assert inputs
    if (!sequence || typeof sequence !== 'string') {
      throw new Error('Sequence must be a non-empty string');
    }
    if (minLength < 1) {
      throw new Error('Minimum length must be positive');
    }

    const code = GeneticCodeTable.GENETIC_CODES.get(codeId);
    if (!code) {
      throw new Error(`Unknown genetic code: ${codeId}`);
    }

    const orfs: Array<{
      start: number;
      end: number;
      frame: number;
      strand: '+' | '-';
      length: number;
      protein: string;
    }> = [];

    // Process both strands
    const sequences = {
      '+': sequence,
      '-': reverseComplement(sequence),
    };

    for (const [strand, seq] of Object.entries(sequences)) {
      const dna = seq.toUpperCase().replace(/U/g, 'T');

      // Check all three frames
      for (let frame = 0; frame < 3; frame++) {
        let inOrf = false;
        let orfStart = -1;
        let orfProtein = '';

        for (let i = frame; i + 2 < dna.length; i += 3) {
          const codon = dna.substring(i, i + 3);
          const aa = code.codons[codon];

          if (aa === undefined || aa === null || aa === '') continue;

          // Check for start codon
          if (!inOrf && code.startCodons.includes(codon)) {
            inOrf = true;
            orfStart = i;
            orfProtein = aa;
          } else if (inOrf) {
            if (aa === '*') {
              // Stop codon found
              if (orfProtein.length >= minLength) {
                orfs.push({
                  start: strand === '+' ? orfStart : sequence.length - i - 3,
                  end: strand === '+' ? i + 2 : sequence.length - orfStart - 1,
                  frame: strand === '+' ? frame : -(frame + 1),
                  strand: strand as '+' | '-',
                  length: orfProtein.length,
                  protein: orfProtein,
                });
              }
              inOrf = false;
              orfProtein = '';
            } else {
              orfProtein += aa;
            }
          }
        }

        // Handle ORF that extends to end of sequence
        if (inOrf && orfProtein.length >= minLength) {
          orfs.push({
            start: strand === '+' ? orfStart : 0,
            end: strand === '+' ? dna.length - 1 : sequence.length - orfStart - 1,
            frame: strand === '+' ? frame : -(frame + 1),
            strand: strand as '+' | '-',
            length: orfProtein.length,
            protein: orfProtein,
          });
        }
      }
    }

    return orfs;
  }

  /**
   * Get genetic code definition by ID
   */
  static getGeneticCode(codeId: GeneticCode): GeneticCodeDefinition | undefined {
    return GeneticCodeTable.GENETIC_CODES.get(codeId);
  }

  /**
   * List all available genetic codes
   */
  static listGeneticCodes(): Array<{
    id: number;
    name: string;
    shortName: string;
  }> {
    return Array.from(GeneticCodeTable.GENETIC_CODES.values()).map((code) => ({
      id: code.id,
      name: code.name,
      shortName: code.shortName,
    }));
  }

  /**
   * Check if a codon is a start codon in the given genetic code
   */
  static isStartCodon(codon: string, codeId: GeneticCode = GeneticCode.STANDARD): boolean {
    const code = GeneticCodeTable.GENETIC_CODES.get(codeId);
    if (!code) return false;

    const normalizedCodon = codon.toUpperCase().replace(/U/g, 'T');
    return code.startCodons.includes(normalizedCodon);
  }

  /**
   * Check if a codon is a stop codon in the given genetic code
   */
  static isStopCodon(codon: string, codeId: GeneticCode = GeneticCode.STANDARD): boolean {
    const code = GeneticCodeTable.GENETIC_CODES.get(codeId);
    if (!code) return false;

    const normalizedCodon = codon.toUpperCase().replace(/U/g, 'T');
    return code.codons[normalizedCodon] === '*';
  }
}
